odoo/tools/translate.py

```diff
@@ -350,6 +350,45 @@ def translate(cr, name, source_type, lang, source=None):
     res = res_trans and res_trans[0] or False
     return res
 
+
+class LazyStrTranslator(object):
+
+    __slots__ = ['_source']
+    def __init__(self, source):
+        self._source = source
+
+    def __repr__(self):
+        _ = GettextAlias()
+        res = self._source
+        cr = None
+        is_new_cr = False
+        try:
+            frame = inspect.currentframe()
+            if frame is None:
+                return self._source
+            frame = frame.f_back
+            if not frame:
+                return self._source
+            lang = _._get_lang(frame)
+            if lang:
+                cr, is_new_cr = _._get_cr(frame)
+                if cr:
+                    # Try to use ir.translation to benefit from global cache if possible
+                    env = odoo.api.Environment(cr, odoo.SUPERUSER_ID, {})
+                    res = env['ir.translation']._get_source(None, ('code','sql_constraint'), lang, self._source)
```

Comment:
Might as well use this opportunity to lint the copied code (E231) ;-)
And perhaps the list of translation types considered by our 2 translation functions `('code', 'sql_constraint')` could be extracted in a constant?